File: iOSReactNativeSample/ios/iOSReactNativeSample/InterstitialAdModule.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <AdWhaleSDK/AdWhaleSDK.h>
#import <React/RCTUtils.h>

@interface InterstitialAdModule : RCTEventEmitter <RCTBridgeModule, AdWhaleInterstitialDelegate>
@property (nonatomic, strong) AdWhaleInterstitialAd *interstitialAd;
@end

@implementation InterstitialAdModule

RCT_EXPORT_MODULE();

- (NSArray<NSString *> *)supportedEvents {
    return @[
        @"onInterstitialAdLoaded",
        @"onInterstitialAdFailedToLoad",
        @"onInterstitialAdFailedToShow",
        @"onInterstitialAdWillPresent",
        @"onInterstitialAdDismissed"
    ];
}

RCT_EXPORT_METHOD(loadAd:(NSString *)adUnitId
                  resolve:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.interstitialAd = [[AdWhaleInterstitialAd alloc] init];
        self.interstitialAd.interstitialDelegate = self;
        [self.interstitialAd load:adUnitId];
        if (resolve) resolve(nil);
    });
}

RCT_EXPORT_METHOD(showAd:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIViewController *rootVC = RCTPresentedViewController();
        if (rootVC) {
            [self.interstitialAd show:rootVC];
            if (resolve) resolve(nil);
        } else {
            if (reject) reject(@"ERROR", @"Root view controller not found", nil);
        }
    });
}

// MARK: - AdWhaleInterstitialDelegate
- (void)adDidReceiveInterstitialAd:(AdWhaleInterstitialAd *)ad {
    self.interstitialAd = ad;
    [self sendEventWithName:@"onInterstitialAdLoaded" body:nil];
}
- (void)adDidFailToReceiveInterstitialAdWithError:(NSError *)error {
    [self sendEventWithName:@"onInterstitialAdFailedToLoad" body:@{@"error": error.localizedDescription ?: @""}];
}
- (void)ad:(AdWhaleInterstitialAd *)ad didFailToPresentInterstitialAdWithError:(NSError *)error {
    [self sendEventWithName:@"onInterstitialAdFailedToShow" body:@{@"error": error.localizedDescription ?: @""}];
}
- (void)adWillPresentInterstitialAd:(AdWhaleInterstitialAd *)ad {
    [self sendEventWithName:@"onInterstitialAdWillPresent" body:nil];
}
- (void)adDidDismissInterstitialAd:(AdWhaleInterstitialAd *)ad {
    [self sendEventWithName:@"onInterstitialAdDismissed" body:nil];
}

@end 

File: iOSExpoSample/ios/iOSExpoSample/InterstitialAdModule.swift
import Foundation
import AdWhaleSDK
import React

@objc(InterstitialAdModule)
class InterstitialAdModule: RCTEventEmitter {
  private var interstitialAd: AdWhaleInterstitialAd?
  
  override func supportedEvents() -> [String]! {
    return [
      "onInterstitialAdLoaded",
      "onInterstitialAdFailedToLoad",
      "onInterstitialAdFailedToShow",
      "onInterstitialAdWillPresent",
      "onInterstitialAdDismissed"
    ]
  }

  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc
  func loadAd(_ adUnitId: String, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      print("[adwhale] [sample Native] 전면 광고 로드 시작")
      self.interstitialAd = AdWhaleInterstitialAd()
      self.interstitialAd?.interstitialDelegate = self
      self.interstitialAd?.load(adUnitId)
      resolve(nil)
    }
  }

  @objc
  func showAd(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
        print("[adwhale] [sample Native] 전면 광고 표시 시도")
        self.interstitialAd?.show(rootViewController)
        resolve(nil)
      } else {
        print("[adwhale] [sample Native] Root view controller를 찾을 수 없음")
        reject("ERROR", "Root view controller not found", nil)
      }
    }
  }
}

extension InterstitialAdModule: AdWhaleInterstitialDelegate {
  func adDidReceiveInterstitialAd(_ ad: AdWhaleInterstitialAd) {
    print("[adwhale] [sample Native] 전면 광고 로드 성공")
    interstitialAd = ad
    sendEvent(withName: "onInterstitialAdLoaded", body: nil)
  }

  func adDidFailToReceiveInterstitialAdWithError(_ error: Error) {
    print("[adwhale] [sample Native] 전면 광고 로드 실패: \(error.localizedDescription)")
    sendEvent(withName: "onInterstitialAdFailedToLoad", body: ["error": error.localizedDescription])
  }

  func ad(_ ad: AdWhaleInterstitialAd, didFailToPresentInterstitialAdWithError error: Error) {
    print("[adwhale] [sample Native] 전면 광고 표시 실패: \(error.localizedDescription)")
    sendEvent(withName: "onInterstitialAdFailedToShow", body: ["error": error.localizedDescription])
  }

  func adWillPresentInterstitialAd(_ ad: AdWhaleInterstitialAd) {
    print("[adwhale] [sample Native] 전면 광고가 표시될 예정")
    sendEvent(withName: "onInterstitialAdWillPresent", body: nil)
  }

  func adDidDismissInterstitialAd(_ ad: AdWhaleInterstitialAd) {
    print("[adwhale] [sample Native] 전면 광고가 닫힘")
    sendEvent(withName: "onInterstitialAdDismissed", body: nil)
  }
} 
 
